File: Search Insert Position.py
# Link: https://leetcode.com/problems/search-insert-position/submissions/

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        
        # Check if target is inside nums
        if target in nums:
            return nums.index(target)
        
        # If target is greater than great element, return list's length + 1
        if target > nums[-1]:
            return len(nums)
        
        # If target is less than smallest element, return 0
        if target < nums[0]:
            return 0
        
        # Search for possible index
        for i in range(len(nums)):
            
            if nums[i] < target and target < nums[i+1]:
                return i + 1

File: Search Insert Position.java
// https://leetcode.com/problems/search-insert-position/

class Solution {
    
    public int binarySearch(int[] nums, int frontIdx, int rearIdx, int target) {
        
        // If indexes overlap, return rearIdx since frontIdx shouldn't be replaced
        // E.g., target = 2 & [1, 3] => [1, 2, 3]
        if (frontIdx > rearIdx) {
            return rearIdx + 1; // <= "+ 1" because question is asking for index with counting start at 1.
        }
        
        // Get middle element
        int midIdx = frontIdx + ((rearIdx - frontIdx) / 2);
        int midElem = nums[midIdx];
        
        // Check if midElem equals to target
        if (midElem == target) {return midIdx;}
        
        // Continue to slice array if target is not found
        if (midElem > target) {
            return binarySearch(nums, frontIdx, midIdx - 1, target);
        }
        else {
            return binarySearch(nums, midIdx + 1, rearIdx, target);
        }
    }
    
    public int searchInsert(int[] nums, int target) {
        return binarySearch(nums, 0, nums.length - 1, target);
    }
}
